staticcheck/lint.go

```diff
@@ -2788,3 +2789,87 @@ func (c *Checker) CheckMissingEnumTypesInDeclaration(j *lint.Job) {
 		ast.Inspect(f, fn)
 	}
 }
+
+func (c *Checker) CheckUnreachableTypeCases(j *lint.Job) {
+	// Check if T subsumes V in a type switch:
+	// T and V are structurally identical interfaces;
+	// interface T is a superset of V;
+	// T is an interface implemented by concrete type V.
+	subsumes := func(T, V types.Type) bool {
+		tIface, ok := T.Underlying().(*types.Interface)
+		if !ok {
+			return false
+		}
+
+		return types.Implements(V, tIface)
+	}
+
+	intersect := func(Ts, Vs []types.Type) (types.Type, types.Type, bool) {
+		for i, T := range Ts {
+			for j, V := range Vs {
+				if subsumes(T, V) {
+					return Ts[i], Vs[j], true
+				}
+			}
+		}
+
+		return nil, nil, false
+	}
+
+	fn := func(node ast.Node) bool {
+		tsStmt, ok := node.(*ast.TypeSwitchStmt)
+		if !ok {
+			return true
+		}
+
+		type ccAndTypes struct {
+			cc    *ast.CaseClause
+			types []types.Type
+		}
+
+		// All asserted types in the order of case clauses.
+		ccs := []ccAndTypes{}
+		for _, stmt := range tsStmt.Body.List {
+			cc, ok := stmt.(*ast.CaseClause)
+			if !ok {
+				continue
+			}
+
+			// Exclude the 'default' case.
+			if len(cc.List) == 0 {
+				continue
+			}
+
+			Ts := []types.Type{}
+			for _, expr := range cc.List {
+				if T := TypeOf(j, expr); T != nil {
```

Comment:
Now that I'm looking at this again, I think not.

My concern was mostly ill–formed and/or partial inputs. Having looked at how the loader and the type checker are initialised, I think we won't ever get in that situation because the check won't even be invoked. Will remove the nil check.

Reply:
That's right. We only check packages with well-formed ASTs and no errors during type-checking.